File: platforms/jvm-default-platform/src/main/java/io/spine/logging/backend/system/SystemClock.java
/*
 * Copyright 2018, The Flogger Authors; 2023, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.spine.logging.backend.system;

import io.spine.logging.jvm.backend.Clock;

import static java.util.concurrent.TimeUnit.MILLISECONDS;

/**
 * Default millisecond precision clock.
 *
 * <p>See class documentation in {@link Clock} for important
 * implementation restrictions.
 *
 * @see <a href="https://rb.gy/veqvb">Original Java code of Google Flogger</a> for historical context.
 */
public final class SystemClock extends Clock {
  private static final SystemClock INSTANCE = new SystemClock();

  // Called during logging platform initialization; MUST NOT call any code that might log.
  public static SystemClock getInstance() {
    return INSTANCE;
  }

  private SystemClock() { }

  @Override
  public long getCurrentTimeNanos() {
    return MILLISECONDS.toNanos(System.currentTimeMillis());
  }

  @Override
  public String toString() {
    return "Default millisecond precision clock";
  }
}

File: platforms/jvm-default-platform/src/main/kotlin/io/spine/logging/backend/system/SystemClock.kt
/*
 * Copyright 2025, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.spine.logging.backend.system

import io.spine.logging.backend.Clock
import java.util.concurrent.TimeUnit.MILLISECONDS

/**
 * Default millisecond precision clock.
 *
 * See class documentation in [Clock] for important implementation restrictions.
 *
 * @see <a href="https://rb.gy/zghcm">Original Java code</a> for historical context.
 */
public class SystemClock private constructor() : Clock() {

    public override fun getCurrentTimeNanos(): Long =
        MILLISECONDS.toNanos(System.currentTimeMillis())

    override fun toString(): String = "Default millisecond precision clock"

    public companion object {
        private val INSTANCE: SystemClock = SystemClock()
        /** Called during logging platform initialization; MUST NOT call any code that might log. */
        @JvmStatic
        public fun getInstance(): SystemClock = INSTANCE
    }
}
